build-all command

Tyler-g-hudson opened the pull request:
Added the Build-All command, which is the omnibus for all of the "build" commands and the semifinal of the docker image creation steps.

Tyler-g-hudson committed:
Fixed a minor CUDA image bug found while building

Tyler-g-hudson committed:
Added build-all command

gmgunter commented on wigwam/cli/build_commands.py:
```diff
@@ -170,6 +173,52 @@ def init_build_parsers(subparsers: argparse._SubParsersAction) -> None:
+        f'have tags fitting "{universal_tag_prefix()}-[TAG]-*". Default: "build"',
+    )
+    parser_build_all.add_argument(
+        "--copy-path",
```
Let's rename this parameter to `--src-path` to be a little more clear that it should be the source prefix path

```suggestion
        "--src-path",
```

gmgunter commented on wigwam/cli/build_commands.py:
```diff
@@ -170,6 +173,52 @@ def init_build_parsers(subparsers: argparse._SubParsersAction) -> None:
+        metavar="FILEPATH",
+        type=str,
+        default=None,
+        help="The path to be copied to the image. If used, no github image will be "
```
```suggestion
        help="The path to the source prefix on the host to be copied to the image. If used, no github image will be "
```

gmgunter commented on wigwam/cli/build_commands.py:
```diff
@@ -170,6 +173,52 @@ def init_build_parsers(subparsers: argparse._SubParsersAction) -> None:
+        type=str,
+        default=None,
+        help="The path to be copied to the image. If used, no github image will be "
+        "copied. Defaults to None.",
```
```suggestion
        help="The path to be copied to the image. If None, instead fetch the source from a remote git archive. Defaults to None.",
```

gmgunter commented on wigwam/cli/build_commands.py:
```diff
@@ -170,6 +173,52 @@ def init_build_parsers(subparsers: argparse._SubParsersAction) -> None:
+        metavar="GIT_ARCHIVE",
+        help='The URL of the Git archive to be fetched. Must be a "tar.gz" file.'
+        "Ignored if --copy-path is used.",
+    )
```
What's the default value for `--archive-url`? What happens if the user doesn't specify either `--copy-path` or `--archive-url`?

gmgunter commented on wigwam/cli/build_commands.py:
```diff
@@ -170,6 +173,52 @@ def init_build_parsers(subparsers: argparse._SubParsersAction) -> None:
+        type=str,
+        metavar="GIT_ARCHIVE",
+        help='The URL of the Git archive to be fetched. Must be a "tar.gz" file.'
+        "Ignored if --copy-path is used.",
```
>Ignored if --copy-path is used.

What happens if the user provides both `--copy-path` AND `--archive-url`? Should we raise an exception to let them know that they've used the interface incorrectly?

One way to handle this is to create a [mutually exclusive parameter group](https://docs.python.org/3/library/argparse.html#argparse.ArgumentParser.add_mutually_exclusive_group) and add both `--copy-path` and `--archive-url` to it. That way, `argparse` will ensure that only one of these two options is selected by the user.

If you use `parser.add_mutually_exclusive_group(required=True)`, then the user is required to specify exactly one of the two options.

gmgunter commented on wigwam/commands.py:
```diff
@@ -303,6 +303,118 @@ def make_distrib(tag: str, base: str, source_tag: str, no_cache: bool = False) -
+    no_cache: bool = False,
+) -> dict[str, Image]:
+    """
+    Fully compiles and builds a Git repo with cmake.
```
```suggestion
    Fully compiles and installs a CMake project.
    
    The source code may be copied into the image from a local directory in the
    host file system or fetched from a remote Git archive.
```

gmgunter commented on wigwam/commands.py:
```diff
@@ -303,6 +303,118 @@ def make_distrib(tag: str, base: str, source_tag: str, no_cache: bool = False) -
+    base: str,
+    copy_path: os.PathLike | None,
+    archive_url: str | None,
+    directory: os.PathLike,
```
```suggestion
    copy_path: str | os.PathLike | None,
    archive_url: str | None,
    directory: str | os.PathLike,
```

gmgunter commented on wigwam/commands.py:
```diff
@@ -303,6 +303,118 @@ def make_distrib(tag: str, base: str, source_tag: str, no_cache: bool = False) -
+def build_all(
+    tag: str,
+    base: str,
+    copy_path: os.PathLike | None,
```
```suggestion
    src_path: os.PathLike | None,
```

gmgunter commented on wigwam/commands.py:
```diff
@@ -303,6 +303,118 @@ def make_distrib(tag: str, base: str, source_tag: str, no_cache: bool = False) -
+        The image tag prefix.
+    base : str
+        The base image tag.
+    copy_path : str
```
```suggestion
    copy_path : path-like or None
```

gmgunter commented on wigwam/commands.py:
```diff
@@ -303,6 +303,118 @@ def make_distrib(tag: str, base: str, source_tag: str, no_cache: bool = False) -
+    base : str
+        The base image tag.
+    copy_path : str
+        The path to a directory to copy to an image.
```
```suggestion
        The path to the source directory on the host to copy to the image.
```

gmgunter commented on wigwam/commands.py:
```diff
@@ -303,6 +303,118 @@ def make_distrib(tag: str, base: str, source_tag: str, no_cache: bool = False) -
+
+    initial_tag: str = ""
+    if is_insert:
+        assert isinstance(copy_path, str)
```
Why does `copy_path` need to be specifically a `str` object? Why not a `pathlib.Path` or some other path-like object?

gmgunter replied:
Is this a good use of `assert`? It looks like we're doing input validation for a public function, so I'd prefer that we raise a `TypeError` instead of an `AssertionError` if this check fails. Remember that `assert` should be used for validating the programmer's own logic and should never actually be tripped by users in operation.

gmgunter commented on wigwam/commands.py:
```diff
@@ -303,6 +303,118 @@ def make_distrib(tag: str, base: str, source_tag: str, no_cache: bool = False) -
+        initial_tag = insert_tag
+    else:
+        git_repo_tag = f"{prefixed_tag}-git-repo"
+        assert archive_url is not None
```
Same here -- I'd prefer that we replace this `assert` with a check that raises a `ValueError` if `archive_url` was `None`.

gmgunter commented on wigwam/commands.py:
```diff
@@ -303,6 +303,118 @@ def make_distrib(tag: str, base: str, source_tag: str, no_cache: bool = False) -
+    if is_insert:
+        assert isinstance(copy_path, str)
+        path_absolute = os.path.abspath(copy_path)
+        if os.path.isdir(copy_path):
+            top_dir = os.path.basename(path_absolute)
+        else:
+            top_dir = os.path.basename(os.path.dirname(path_absolute))
```
I'm not a fan of this behavior. If `copy_path` is not a directory, IMO we should raise an error here instead of getting the parent directory using `os.path.dirname`.

gmgunter commented on wigwam/commands.py:
```diff
@@ -303,6 +303,118 @@ def make_distrib(tag: str, base: str, source_tag: str, no_cache: bool = False) -
+        insert_image = copy_dir(
+            base=prefixed_base_tag,
+            tag=insert_tag,
+            directory=directory,
+            target_path=copy_path,
```
Shouldn't `directory` and `copy_path` be swapped here?

According to [the docstring for `copy_dir`](https://github.com/isce3-testing/image-sandbox/blob/de9168c5994d3400f9e977e1279b9710650f68e5/wigwam/commands.py#L91-L121), `directory` should be the source directory on the host to be copied, and `target_path` is the destination path in the image.

But according to [the docstring of this function (`build_all`)](https://github.com/isce3-testing/image-sandbox/blob/de9168c5994d3400f9e977e1279b9710650f68e5/wigwam/commands.py#L316-L345), `directory` is the path of the destination directory inside the image, and `copy_path` is the source directory on the host.

gmgunter replied:
Maybe we should take this as an indication that we should come up with more descriptive names for these variables, e.g. `host_dir` and `image_dir`, or `src_path` and `dst_path`.

gmgunter reviewed (commented):
Thanks! Let's add a unit test that checks that you can `import isce3` and `import nisar` in the resulting image.

Tyler-g-hudson committed:
PR Responses

+ Added type annotations to CLI to help dev environments follow the code
~ Change "directory" and "copy-path" arguments to "dst-path", "src-path"
~ Improved type checking and resulting behaviors

Tyler-g-hudson committed:
Added a use_prefix option for dropin sessions
